src/sources/datadog/agent/traces.rs

```diff
@@ -0,0 +1,219 @@
+                                query_params.dd_api_key,
+                            ),
+                            reported_language.as_ref(),
+                            &source,
+                        )
+                        .map_err(|error| {
+                            ErrorMessage::new(
+                                StatusCode::UNPROCESSABLE_ENTITY,
+                                format!("Error decoding Datadog traces: {:?}", error),
+                            )
+                        })
+                    });
+                if multiple_outputs {
+                    handle_request(events, acknowledgements, out.clone(), Some(agent::TRACES))
+                } else {
+                    handle_request(events, acknowledgements, out.clone(), None)
+                }
+            },
+        )
+        .boxed()
+}
+
+fn build_stats_filter() -> BoxedFilter<(Response,)> {
+    warp::post()
+        .and(path!("api" / "v0.2" / "stats" / ..))
+        .and_then(|| {
+            warn!(
+                message = "/api/v0.2/stats route is yet not supported.",
+                internal_log_rate_secs = 60
+            );
+            let response: Result<Response, Rejection> = Ok(warp::reply().into_response());
+            future::ready(response)
+        })
+        .boxed()
+}
+
+fn handle_dd_trace_payload(
+    frame: Bytes,
+    api_key: Option<Arc<str>>,
+    lang: Option<&String>,
+    source: &DatadogAgentSource,
+) -> crate::Result<Vec<Event>> {
+    let decoded_payload = dd_proto::TracePayload::decode(frame)?;
+    let env = decoded_payload.env;
+    let hostname = decoded_payload.host_name;
+    let trace_events: Vec<Event> =
+    // Each traces is mapped to one event...
+    decoded_payload
+        .traces
+        .iter()
+        .map(|dd_trace| convert_dd_trace(dd_trace, env.clone(), hostname.clone(), source))
+        //... and each APM event is also mapped into its own event
+        .chain(decoded_payload.transactions.iter().map(|s| {
+            let mut trace_event = TraceEvent::from(convert_span(s));
+            trace_event.insert(
+                source.log_schema_source_type_key,
+                Bytes::from("datadog_agent"),
+            );
+            trace_event.insert(source.log_schema_host_key, hostname.clone());
+            trace_event.insert("env", env.clone());
+            trace_event
+        }))
+        .map(|mut trace_event| {
+            if let Some(k) = &api_key {
+                trace_event
+                    .metadata_mut()
+                    .set_datadog_api_key(Some(Arc::clone(k)));
+            }
+            if let Some(lang) = lang {
+                trace_event.insert("language", lang.clone());
+            }
+            trace_event.insert("payload_version", "v1".to_string());
+            Event::Trace(trace_event)
+        })
+        .collect();
+    emit!(&EventsReceived {
+        byte_size: trace_events.size_of(),
+        count: trace_events.len(),
+    });
```

Comment:
We should emit this _before_ any enrichment. `EventsSent` will include the enriched event size.